src/overlays/actors/ovl_En_Mt_tag/z_en_mt_tag.c

```diff
@@ -35,42 +40,475 @@ const ActorInit En_Mt_tag_InitVars = {
+    s32 hasSetCurrentCheckpointOnce = false;
+    f32 minLineLengthSq = 0.0f;
+    s32 sceneExitIndex;
+    f32 perpendicularPointX;
+    f32 perpendicularPointZ;
+    f32 lineLenSq;
+    s32 checkpointIterator;
+
+    // The Goron Racetrack is configured such that the sceneExitIndex for any given floor polygon
+    // gradually increases as you move forward through the racetrack.
+    sceneExitIndex = SurfaceType_GetSceneExitIndex(&globalCtx->colCtx, actor->floorPoly, actor->floorBgId);
+    if ((sceneExitIndex < 4) || (sceneExitIndex >= 19)) {
+        //! @bug - upcomingCheckpoint is not initialized here
+        return -1;
+    }
+
+    checkpointIterator = sStartingCheckpointPerSceneExitIndex[sceneExitIndex];
+
+    // Iterates through all possible checkpoints that are associated with this sceneExitIndex.
+    do {
+        if ((Math3D_PointDistToLine2D(
+                actor->world.pos.x, actor->world.pos.z, (&sCheckpointPositions[checkpointIterator])[-1].x,
+                (&sCheckpointPositions[checkpointIterator])[-1].z, (&sCheckpointPositions[checkpointIterator])[1].x,
+                (&sCheckpointPositions[checkpointIterator])[1].z, &perpendicularPointX, &perpendicularPointZ,
+                &lineLenSq)) &&
+            (!hasSetCurrentCheckpointOnce || ((curentCheckpoint + 1) == checkpointIterator) ||
+             (lineLenSq < minLineLengthSq))) {
+            minLineLengthSq = lineLenSq;
+            curentCheckpoint = checkpointIterator;
+            *outPerpendicularPointX = perpendicularPointX;
+            *outPerpendicularPointZ = perpendicularPointZ;
+            hasSetCurrentCheckpointOnce = true;
+        }
+        checkpointIterator++;
+    } while (checkpointIterator < sStartingCheckpointPerSceneExitIndex[sceneExitIndex + 1]);
+
+    *upcomingCheckpoint = curentCheckpoint + 1;
+    return curentCheckpoint;
+}
+
+/**
+ * Returns true if the player is almost certainly going to lose the race.
+ * Specifically, it checks if the player's current checkpoint is 24 or more
+ * checkpoints behind the leading racer. This value was probably chosen because
+ * falling off the wooden bridge in the middle of the track can set the player
+ * back up to 23 checkpoints.
+ *
+ * This function also has the side effect of updating the number of checkpoints
+ * ahead of the player each Race Goron is.
+ */
+s32 EnMttag_UpdateCheckpoints(EnMttag* this, GlobalContext* globalCtx) {
+    Player* player = GET_PLAYER(globalCtx);
+    EnRg* rg;
+    s32 currentCheckpoints[5];
+    s32 upcomingCheckpoints[5];
+    f32 perpendicularPointsX[5];
+    f32 perpendicularPointsZ[5];
+    s32 highestCurrentCheckpoint;
+    s32 i;
+    s32 ret = false;
+
+    highestCurrentCheckpoint = -1;
+    currentCheckpoints[0] = EnMttag_GetCurrentCheckpoint(&player->actor, globalCtx, &upcomingCheckpoints[0],
+                                                         &perpendicularPointsX[0], &perpendicularPointsZ[0]);
+    for (i = 1; i < 5; i++) {
```

Comment:
I feel like this should be using an ARRAY_COUNT, but it is kinda hard to say which symbol use.